### PR TITLE
Added IPv6 support for the ipfailover keepalived image. (bz1459960)

### PR DESCRIPTION
**Pull request by rmedinaday**
- Added IPv6 Address support
 - Added IPv6 Address Range Support
 - Added IPv6 Address Validation
 - Added IPv4 Address Validation
 - Added relevant test cases

### CODE REVIEW

File: images/ipfailover/keepalived/lib/utils.sh [hunk 1]
```diff
@@ -21,6 +21,154 @@ function scrub() {
...
+
+  if [[ ${1} =~ ${IPv4_SHAPE} ]]; then
+    local is_valid=0
+    for i in $(echo ${1} | tr "." " "); do
+      if [ $i -gt 255 ]; then
+        local is_valid=1
+      fi
+    done
+  else
+    local is_valid=1
```
Comment (bmeng) [inline images/ipfailover/keepalived/lib/utils.sh]: When it says `is_valid=1`, I think it should be `is_valid=true`. But it is invalid actually. 
A little reverse my brain.

Comment (rmedinaday) [inline images/ipfailover/keepalived/lib/utils.sh]: Yes, that's due to "true" being 0 in bash, so is_valid=1 is actually is_valid=false. As we're dealing with function returns, not echoes, I found it best to stick to the convention.

File: images/ipfailover/keepalived/lib/utils.sh [hunk 2]
```diff
@@ -21,6 +21,154 @@ function scrub() {
...
+#     validate_ipv4 192.0.2 
+#         # -> 1 
+#
+function validate_ipv4() {
+  local IPv4_GROUP="[0-2]?[0-9]{1,2}"
+  local IPv4_SHAPE="^(${IPv4_GROUP}\.){3,3}(${IPv4_GROUP})$"
+
+  if [[ ${1} =~ ${IPv4_SHAPE} ]]; then
+    local is_valid=0
+    for i in $(echo ${1} | tr "." " "); do
```
Comment (php-coder) [inline images/ipfailover/keepalived/lib/utils.sh]: If you're using locals, why not define `i` as `local`?

**Comment by knobunc** [comment]
[test][testextended][extended: networking]

**Comment by imcsk8** [comment]
code looks ok [LGTM]

**Comment by knobunc** [comment]
@openshift/networking PTAL

**Comment by knobunc** [comment]
@rmedinaday Squash the commits please and I will merge this on Monday when the gates re-open.  Thanks

**Comment by rmedinaday** [comment]
Thanks @knobunc, Sorry to be such a nuisance, but this is my first contribution so I'm a bit at a loss as to how to squash the commits without breaking everything... From what I understand, I have to do it on my forked repo and it'll refresh it here, right?

**Comment by knobunc** [comment]
[merge]

**Comment by deads2k** [comment]
re[test]

**Comment by eparis** [comment]
[merge]
[test]
Both were flakes

**Comment by knobunc** [comment]
[merge] Last flaked on a provisioning error and a devicemapper error https://github.com/openshift/origin/issues/13271

**Comment by smarterclayton** [comment]
[test]

**Comment by smarterclayton** [comment]
[test]

On Jun 27, 2017, at 11:15 PM, OpenShift Bot <notifications@github.com>
wrote:

continuous-integration/openshift-jenkins/merge FAILURE (
https://ci.openshift.redhat.com/jenkins/job/merge_pull_request_origin/1156/)
(Base Commit: a67ac87
<https://github.com/openshift/origin/commit/a67ac87be7b0bc35d0f2b79cfff4d15f52f51178>)
(PR Branch Commit: c5f949b
<https://github.com/openshift/origin/commit/c5f949bfc208f5594f8403be47d3d461685ba3e9>
)

—
You are receiving this because you commented.
Reply to this email directly, view it on GitHub
<https://github.com/openshift/origin/pull/14527#issuecomment-311545253>,
or mute
the thread

**Comment by knobunc** [comment]
[merge] Last merge flaked on https://github.com/openshift/origin/issues/14043

**Comment by knobunc** [comment]
[test] [merge]

**Comment by eparis** [comment]
[merge]

File: images/ipfailover/keepalived/lib/utils.sh [hunk 3]
```diff
@@ -36,26 +188,26 @@ function scrub() {
 function expand_ip_ranges() {
   local vips=${1:-""}
   local expandedset=()
+  local iprange
+  local newip
 
-  for iprange in $(echo "$vips" | sed 's/[^0-9\.\,-]//g' | tr "," " "); do
```
Comment (rmedinaday) [inline images/ipfailover/keepalived/lib/utils.sh]: Not sure why it was that way before, as the sed s command uses single quotes (so it's basically a literal and needs no escaping, save for single quotes), and ',' isn't a sed regexp special character. '\.' on the other hand does need the quote for the regexp to be meaningful, so I'm guessing it was more of a "it won't hurt if we add it" kind of thing... If someone wants to shed some light on it, by all means do!

**Pull request merged**